applogic/src/nse/mod.rs

```diff
@@ -99,3 +63,88 @@ pub extern "C" fn free_string(s: *mut c_char) {
         let _ = CString::from_raw(s);
     }
 }
+
+fn error_batch(e: String) -> NotificationBatch {
+    NotificationBatch {
+        badge_count: 0,
+        removals: Vec::new(),
+        additions: vec![NotificationContent {
+            identifier: "".to_string(),
+            title: "Error".to_string(),
+            body: e,
+            data: "".to_string(),
+        }],
+    }
+}
+
+/// Wraps with a tokio runtime to block on the async functions
+fn retrieve_messages_sync(path: String) -> NotificationBatch {
+    match Builder::new_multi_thread()
+        .thread_name("nse-thread")
+        .enable_all()
+        .build()
+    {
+        Ok(runtime) => {
+            match panic::catch_unwind(AssertUnwindSafe(|| {
+                runtime.block_on(async { retrieve_messages(path).await })
+            })) {
+                Ok(batch) => batch,
+                Err(_) => match panic::catch_unwind(AssertUnwindSafe(|| drop(runtime))) {
+                    Ok(_) => {
+                        let e = "Failed to execute block_on";
+                        log::error!("{}", e);
+                        error_batch(e.to_string())
+                    }
+                    Err(_) => {
+                        let e = "Failed to drop runtime";
+                        log::error!("{}", e);
+                        error_batch(e.to_string())
```

Comment:
This whole structure looks a bit weird, but the only thing I think we can do here is pull the last two lines out that are the same in the `Ok(_)` and the `Err(_)` branch.